builtin/credential/ldap/backend.go

```diff
@@ -122,6 +122,17 @@ func (b *backend) Login(req *logical.Request, username string, password string)
 		return nil, logical.ErrorResponse(fmt.Sprintf("LDAP bind failed: %v", err)), nil
 	}
 
+	// We re-bind to the BindDN if it's defined because we assume
+	// the BindDN should be the one to search, not the user logging in.
+	if cfg.DiscoverDN || (cfg.BindDN != "" && cfg.BindPassword != "") {
+		if err := c.Bind(cfg.BindDN, cfg.BindPassword); err != nil {
+			return nil, logical.ErrorResponse(err.Error()), nil
+		}
+		if b.Logger().IsDebug() {
+			b.Logger().Debug("auth/ldap: Re-Bound to original BindDN")
+		}
+	}
+
 	userDN, err := b.getUserDN(cfg, c, bindDN)
```

Comment:
Between `bindDN` and `cfg.BindDN`, following the flow is a bit challenging. Perhaps we should rename the `bindDN` local variable to something like `userBindDN`, to differentiate it from the _service_ bind DN - `cfg.BindDN` ?

Also note `userBindDN` would not necessarily equal `userDN` - userDN is the canonical representation of the user, whereas `userBindDN` could in fact be using a different mechanism such as user principal name (`user@domain`).